ompi/mca/osc/ucx/osc_ucx_comm.c

```diff
@@ -1212,3 +1279,388 @@ int ompi_osc_ucx_rget_accumulate(const void *origin_addr, int origin_count,
+                    struct ompi_datatype_t *stage_dt, int target, ptrdiff_t target_disp,
+                    int target_count, struct ompi_datatype_t *target_dt, struct ompi_op_t
+                    *op, struct ompi_win_t *win, bool lock_acquired, const void
+                    *origin_addr, int origin_count, struct ompi_datatype_t *origin_dt, bool is_put,
+                    int phase, int acc_type) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    ucp_ep_h *ep;
+    OSC_UCX_GET_DEFAULT_EP(ep, module, target);
+    opal_common_ucx_wpmem_t *mem = module->mem;
+    uint64_t remote_addr = (module->state_addrs[target]) + OSC_UCX_STATE_REQ_FLAG_OFFSET;
+    ompi_osc_ucx_request_t *ucx_req = NULL;
+    bool sync_check;
+    int ret = OMPI_SUCCESS;
+    CHECK_DYNAMIC_WIN(remote_addr, module, target, ret, true);
+
+    if (acc_type != NONE) {
+        OMPI_OSC_UCX_REQUEST_ALLOC(win, ucx_req);
+        assert(NULL != ucx_req);
+        ucx_req->acc.op = op;
+        ucx_req->acc.acc_type = acc_type;
+        ucx_req->acc.phase = phase;
+        ucx_req->acc.module = module;
+        ucx_req->acc.target = target;
+        ucx_req->acc.lock_acquired = lock_acquired;
+        ucx_req->acc.win = win;
+        ucx_req->acc.origin_addr = origin_addr;
+        ucx_req->acc.origin_count = origin_count;
+        if (origin_dt != NULL) {
+            ucx_req->acc.origin_dt = origin_dt;
+            if (!ompi_datatype_is_predefined(origin_dt)) {
+                OBJ_RETAIN(ucx_req->acc.origin_dt);
+            }
+        }
+        ucx_req->acc.stage_addr = stage_addr;
+        ucx_req->acc.stage_count = stage_count;
+        if (stage_dt != NULL) {
+            ucx_req->acc.stage_dt = stage_dt;
+            if (!ompi_datatype_is_predefined(stage_dt)) {
+                OBJ_RETAIN(ucx_req->acc.stage_dt);
+            }
+        }
+        ucx_req->acc.target = target;
+        if (target_dt != NULL) {
+            ucx_req->acc.target_dt = target_dt;
+            if (!ompi_datatype_is_predefined(target_dt)) {
+                OBJ_RETAIN(ucx_req->acc.target_dt);
+            }
+        }
+        ucx_req->acc.target_disp = target_disp;
+        ucx_req->acc.target_count = target_count;
+        ucx_req->acc.free_ptr = NULL;
+    }
+    sync_check = module->skip_sync_check;
+    module->skip_sync_check = true; /* we already hold the acc lock, so no need for sync check*/
+
+    if (is_put) {
+        ret = ompi_osc_ucx_put(origin_addr, origin_count, origin_dt, target, target_disp,
+                               target_count, target_dt, win);
+    } else {
+        ret = ompi_osc_ucx_get(stage_addr, stage_count, stage_dt, target, target_disp,
+                target_count, target_dt, win);
+    }
+    if (ret != OMPI_SUCCESS) {
+        return ret;
+    }
+
+    module->skip_sync_check = sync_check;
+    if (acc_type != NONE) {
+        OSC_UCX_INCREMENT_OUTSTANDING_NB_OPS();
```

Comment:
Is this thread-safe? The increment is non-atomic and there doesn't seem to be a lock around this code path (or I'm not seeing it?)